File: C/codigo.c
#include<stdio.h>
#include<string.h>
#include<unistd.h>
#include<stdlib.h>
#include<time.h>

int main(int argc, char** argv){
    const char* GPIOS[8] = {"26","19","6","5","9","10","22","27"};

    if(argc <= 1){
        printf("No hay comandos\n");
        printf("los comandos a utilizar es config, valor, cerrar\n");
        return 1;
    }

    if(strcmp(argv[1],"config") == 0){
        char addres[33];
        FILE* exp = fopen("/sys/class/gpio/export","a");
        FILE* dir = NULL;

        if (exp == NULL) {
            fprintf(stdout,"Error al abrir /sys/class/gpio/export");
            return 1;
        }

        for(int i=0;i<8;i++){
            printf("Exporanto GPIO numero %s\n",GPIOS[i]);
            fprintf(exp,"%s",GPIOS[i]);
            fflush(exp); 
            sleep(1);

            strcpy(addres,"/sys/class/gpio/gpio");
            strcat(addres,GPIOS[i]);
            strcat(addres,"/direction");

            dir = fopen(addres,"a");
            fprintf(dir,"in");
            fclose(dir);
        }
        fclose(exp);
    } else if(strcmp(argv[1],"cerrar") == 0){
        FILE* unexp = fopen("/sys/class/gpio/unexport","a");
        for(int i=0;i<8;i++){
            printf("Cerrando el GPIO %s\n",GPIOS[i]);
            fprintf(unexp,"%s",GPIOS[i]);
        }
        fclose(unexp);
    }else if(strcmp(argv[1],"valor") == 0){
        struct timespec ts;
        FILE* pin = NULL;
        FILE* amplitudes = NULL;
        FILE* times = NULL;
        char bits[8][29];
        int numero = 0;
        char* amplitud = (char*)calloc(10,sizeof(char));
        char* hhmmss = (char*)calloc(9,sizeof(char));
        char* time = (char*)calloc(20,sizeof(char));

        for(int i=0;i<8;i++)
            snprintf(bits[i],sizeof(bits[i]),"/sys/class/gpio/gpio%s/value",GPIOS[i]);

        for(int i=0;i<100;i++){
	    numero = 0;
            amplitudes = fopen("amplitudes.txt","a");
            times = fopen("times.txt","a");
            // get number
	    usleep(100000);
            for(int i=0;i<8;i++){
                pin = fopen(bits[i],"r");
                fscanf(pin,"%s",amplitud);
                fclose(pin);
                numero += atoi(amplitud);
            }
            // get time
            clock_gettime(CLOCK_REALTIME, &ts);
            strftime(hhmmss, 9, "%X", localtime(&ts.tv_sec));
	    snprintf(time,20,"%s:%09ld", hhmmss, ts.tv_nsec);
            
            fprintf(times,"%s\n",time);
            fprintf(amplitudes,"%d\n",numero);
        } 
        fclose(amplitudes);
        fclose(times);
        free(amplitud); 
        free(time); 
        free(hhmmss); 
    }
}

File: Python/codigo.py
import os
import time
from datetime import datetime

GPIOS = ["26","19","6","5","9","10","22","27"]

def export_gpios():
    with open("/sys/class/gpio/export", "a") as exp:
        for gpio in GPIOS:
            print(f"Exportando GPIO número {gpio}")
            exp.write(gpio)
            exp.flush()
            time.sleep(1)
            direction_path = f"/sys/class/gpio/gpio{gpio}/direction"
            with open(direction_path, "w") as dir_file:
                dir_file.write("in")

def unexport_gpios():
    with open("/sys/class/gpio/unexport", "w") as un:
        for gpio in GPIOS:
            print(f"Cerrando el GPIO {gpio}")
            un.write(f"{gpio}")
            un.flush()

def read_values():
    gpio_paths = [f"/sys/class/gpio/gpio{gpio}/value" for gpio in GPIOS]
    amplitudes_file = open("amplitudes.txt", "a")
    times_file = open("times.txt", "a")

    for _ in range(100):
        total_value = 0
        # Lee los valores de los pines GPIO
        for gpio_path in gpio_paths:
            with open(gpio_path, "r") as pin:
                total_value += int(pin.read().strip())

        # Obtén la hora actual con nanosegundos
        now = datetime.now()
        current_time = now.strftime("%H:%M:%S")
        nanoseconds = f"{now.microsecond * 1000:09d}"
        timestamp = f"{current_time}:{nanoseconds}"

        # Guarda en los archivos
        times_file.write(f"{timestamp}\n")
        amplitudes_file.write(f"{total_value}\n")

        time.sleep(0.1)  # Pausa de 100 ms

    amplitudes_file.close()
    times_file.close()

if __name__ == "__main__":
    import sys

    if len(sys.argv) <= 1:
        print("No hay comandos")
        print("Los comandos a utilizar son: config, valor, cerrar")
        sys.exit(1)

    command = sys.argv[1]

    if command == "config":
        export_gpios()
    elif command == "cerrar":
        unexport_gpios()
    elif command == "valor":
        read_values()
    else:
        print("Comando no reconocido")
